### PR TITLE
Pandas assignment - Lotfi Kobrosly

### PR DESCRIPTION
**Pull request by LotfiKobrosly**

### CODE REVIEW

**Comment by tomMoral** [comment]
If you are finished, you can sign up for this RAMP challenge: https://ramp.studio/events/brevet_datacamp2020
This is what we are going to do this afternoon. The goal is to try and solve the challenge.
You need to click on the blue button: `Join the event` and I need to approve your participation so send me a message once you have done this.
You can then read the starting kit and go to the left menu to submit your solutions in the sandbox.

File: data/.~lock.departments.csv# [hunk 1]
```diff
@@ -0,0 +1 @@
+,lotfi,lotfi-Lenovo-ideapad-330-15IKB,10.11.2020 00:20,file:///home/lotfi/.config/libreoffice/4;
```
Comment (tomMoral) [inline data/.~lock.departments.csv#:1]: You need to be carefull and not commit extra files.

Can you remove them?

File: pandas_questions.py [hunk 2]
```diff
@@ -9,15 +9,22 @@
...
-    departments = pd.DataFrame({})
+    referendum = pd.DataFrame(pd.read_csv("./data/referendum.csv", sep=";"))
+    regions = pd.DataFrame(pd.read_csv("./data/regions.csv"))
+    departments = pd.DataFrame(pd.read_csv("./data/departments.csv"))
+    # Column names
+    # regions: id, code, name, slug
+    # departments: id, region_code, code, name, slug
+    # referendum: Department code, Department name, Town code,
+    #             Town name, Registered, Abstentions, Null,
+    #             Choice A, Choice B
```
Comment (tomMoral) [inline pandas_questions.py:27]: Clean up your debugging comments.

File: pandas_questions.py [hunk 3]
```diff
@@ -27,29 +34,101 @@ def merge_regions_and_departments(regions, departments):
...
+
+    Returns
+    -------
+    df : pandas.DataFrame
+        merged dataframes
+
+    Raises
+    ------
+    None
+
```
Comment (tomMoral) [inline pandas_questions.py:54]: Nice one (the other did not do this)! It is not mandatory but you can remove `Raises` when you don't raise error.

File: pandas_questions.py [hunk 4]
```diff
@@ -27,29 +34,101 @@ def merge_regions_and_departments(regions, departments):
...
     """
+    df = pd.DataFrame({
+        "code_reg": departments["region_code"],
+        "name_reg": ["" for _ in range(departments.shape[0])],
+        "code_dep": departments["code"],
+        "name_dep": departments["name"]
+    })
+    for i, row in regions.iterrows():
+        mask = df["code_reg"] == row["code"]
+        df.loc[mask, "name_reg"] = row["name"]
```
Comment (tomMoral) [inline pandas_questions.py:64]: Could you try to do this with the `merge` method?
This for loop is quite inefficient I think

File: pandas_questions.py [hunk 5]
```diff
@@ -27,29 +34,101 @@ def merge_regions_and_departments(regions, departments):
...
+        dep_code = row["code_dep"]
+        try:
+            dep_code = str(int(dep_code))
+        except:
+            pass
+        mask = (df["Department code"] == dep_code)
+        df["code_dep"] = df["code_dep"].mask(mask, row["code_dep"]).to_numpy()
+        df["name_dep"] = df["name_dep"].mask(mask, row["name_dep"]).to_numpy()
+        df["code_reg"] = df["code_reg"].mask(mask, row["code_reg"]).to_numpy()
+        df["name_reg"] = df["name_reg"].mask(mask, row["name_reg"]).to_numpy()
```
Comment (tomMoral) [inline pandas_questions.py:106]: Same, use `pd.merge`

**Review by tomMoral — CHANGES_REQUESTED** [review]
Can you update your PR to remove extra files and use `merge` to merge the dataframe?

File: data/.~lock.departments.csv# (hunk 1 above)
Comment (LotfiKobrosly) [reply data/.~lock.departments.csv#:1]: I cannot see these files in the directory (even with `ls -lart` ). Could you suggest me how I can fix it please?

Comment (LotfiKobrosly) [reply data/.~lock.departments.csv#:1]: I guess it is due to me opening the `.csv` files using LibreOffice

Comment (agramfort) [reply data/.~lock.departments.csv#:1]: `git rm .~lock.departments.csv#` would remove one file

**Comment by LotfiKobrosly** [comment]
Changes added thank you !

File: pandas_questions.py [hunk 6]
```diff
@@ -60,9 +130,31 @@ def plot_referendum_map(referendum_result_by_regions):
...
     """
+    region_map = gpd.read_file("./data/regions.geojson")
 
-    return gpd.GeoDataFrame({})
+    df = {"geometry": list()}
+    for col in referendum_result_by_regions.columns:
+        df[col] = referendum_result_by_regions[col]
+    for i, row in referendum_result_by_regions.iterrows():
+        mask = region_map["nom"] == row["name_reg"]
+        df["geometry"].append(region_map.loc[mask]["geometry"].values[0])
```
Comment (tomMoral) [inline pandas_questions.py:151]: You can also use a merge here.

**Review by tomMoral — CHANGES_REQUESTED** [review]
Still a few comments on the PR.

**Review by tomMoral — CHANGES_REQUESTED** [review]
Look mostly good, except for the extra `pd.DataFrame` in `load_data`.

File: pandas_questions.py [hunk 7]
```diff
@@ -27,29 +28,98 @@ def merge_regions_and_departments(regions, departments):
...
+    df : pandas.DataFrame
+        merged dataframes
+
     """
+    df = pd.merge(
+        departments,
+        regions,
+        left_on='region_code',
+        right_on='code',
+        suffixes=['_dep', '_reg'])
```
[the remark this answers is not in the document]
Comment (LotfiKobrosly) [reply pandas_questions.py]: I actually usually do that :)

**Comment by LotfiKobrosly** [comment]
All done thank you for your remarks!

**Review by tomMoral — APPROVED** [review]
LGTM!!

**Pull request closed without merging**